Adding UI dropdown for model selection

hegdes-usc opened the pull request:
# Description

UI dropdown prototype for model selection.

Note that as of this PR, the additional models are not supported without setting API Keys as env variables as in [https://github.com/mito-ds/mito/pull/1644](url)

https://github.com/user-attachments/assets/6d453890-92bd-4c85-ac06-4e43c4e94551


# Testing

### Note: This feature requires additional work in order to make it work without API keys as discussed with @aarondr77. Hence this will serve as the ground work for the same

- [ ] I have tested this on real data that is reasonable and large
- [ ] If I changed the interaction with JupyterLab, I tested that it does not break other programs (like VS Code), and tested that it works "multiple times" in the same notebook.

aarondr77 commented on mito-ai/mito_ai/constants.py:
```diff
@@ -3,6 +3,12 @@
 
 import os
 
+# Active model selections (updated by UI)
```
We'd be able to delete these

aarondr77 commented on mito-ai/mito_ai/constants.py:
```diff
@@ -20,8 +26,60 @@
+AZURE_OPENAI_MODEL = os.environ.get("AZURE_OPENAI_MODEL")
+
+
+def set_active_model(model):
```
We could delete all of this too

aarondr77 commented on mito-ai/mito_ai/providers.py:
```diff
@@ -187,8 +219,14 @@ def capabilities(self) -> AICapabilities:
+        current_model = self._resolve_model()
+
+        # If model has changed or client is closed, rebuild the client
+        if self._active_model != current_model or not self._async_client or self._async_client.is_closed():
```
Nice good catch

aarondr77 commented on mito-ai/mito_ai/providers.py:
```diff
@@ -198,24 +236,27 @@ def key_type(self) -> str:
 
         return MITO_SERVER_KEY
 
     def _build_openai_client(self) -> Optional[Union[openai.AsyncOpenAI, openai.AsyncAzureOpenAI]]:
```
I think we'd update this function to take in the `active_model` as a parameter and then we would do the following:

```python

if get_active_model_provider(active_model) == 'claude' and  constants.CLAUDE_MODEL and constants.CLAUDE_API_KEY:
             base_url = constants.CLAUDE_BASE_URL
             llm_api_key = constants.CLAUDE_API_KEY
             self.log.debug(f"Using Claude with model: {active_claude}")


# Then the same thing for the other providers ...
```

aarondr77 reviewed (changes requested):
Nice work on this. It's all coming together and I'm excited to get this released to folks. 

I think we can pretty dramatically simplify this PR if we make one design change: There is only ever one active model and that model is always whatever the user selects in the UI. We can then:

1. Just store the active model in the provider.py class 
2. When a completion message is sent, we can check if the user has environment variables set so we can build a provider for that model. If they do, we use their key to use that model. If they don't, we use the Mito Server instead. 


TLDR: Unless I'm thinking about it incorrectly, there's no need to have an active model per model provider. We can just have one active model and either use the user's key to execute it or use the Mito server.

aarondr77 commented:
Feel free to ignore the Mito AI Frontend Playwright failing tests. Those are flakey and fail often, but let's fix up the rest of the tests before merging

hegdes-usc commented:
Hey @aarondr77 I have addressed all the review comments and have pushed a new commit with other changes as well. As discussed on slack

Different models return different types of outputs, and in order to handle those I have made a few changes in the way chunks are processed along with the changes you requested.

There are still a few issues in the way these models return outputs for example some responses from the claude models do not enclose the code snippets in the right format, which can block the user from overriding the current cell with the output. 

It is sure that there will be a lot of these edge cases which would require a much more thorough approach specific to each model provider.

aarondr77 commented on mito-ai/mito_ai/providers.py:
```diff
@@ -353,60 +399,89 @@ async def stream_completions(
+                            done=False,
+                        ))
+
+                    if is_finished:
```
What's the purpose of emitting the full final response at the end? I think the stream handler on the frontend doesn't need this information.

hegdes-usc replied:
This is to handle a case with the gemini flash model. It responds so quickly, that it returns the whole response in one chunk. 

Sidenote: We should use one of the gemini flash models to do inline completions.

aarondr77 commented on mito-ai/mito_ai/constants.py:
```diff
@@ -20,8 +23,53 @@
+    if ACTIVE_MODEL:
+        return ACTIVE_MODEL
+
+    # Default to environment variables if no active model
+    for model in [CLAUDE_MODEL, GEMINI_MODEL, OLLAMA_MODEL]:
+        if model:
+            return model
```
There should always be an `active model` I think since the UI is always going to display a model. So I think we can remove this

hegdes-usc replied:
This function is needed during initial startup before the UI fully loads else it throws an error since this appears to load first. 

This also helps if there's a bug in the UI model selection as well. This just makes it more robust to unexpected issues.

aarondr77 reviewed (changes requested):
Nice work. I like these simplifications that you made. 

Yeah, so the problem with Claude is [Claude's OpenAI SDK compatibility does not support the response_format parameter](https://docs.anthropic.com/en/api/openai-sdk#simple-fields). 

I think we should do the following since adding the model UI selector + adding support for all OpenAI models and Gemini models is going to be a really exciting and useful feature for people:
1. Let's keep everything as is, but comment out the anthropic models from the UI selector. That way, users can never end up with an anthropic model for now. 
2. Let's get this PR merged in. 
3. Let me invite you to our lambda function repo so you can get the code implemented on the mito server for users to be able to select their model even when they are using our server. This is going to be the most common user flow. I think we'll need to copy over the `get_model_type` function so we can choose which API key / endpoint to use. 

Then let's touch base on some next steps. A few things that we might want to do after that:
1. Move the AI model selector to the bottom like cursor. I showed the existing implementation to a few users and they shared this feedback that at the bottom is more expected since it's where the rest of their chat configuration options are. Before we do this, we can make a tiny mockup so we can align on the design.
2. Currently, we are setting the provider for all types of AI completions with this UI. However, the model that is best for users for chat is likely not the right model for inline code completions. For chat, thinking models might make sense. For inline code completion, speed is crucial. So one thing I could imagine doing is: when the user selects the model, we use that for the chat, and then the use the max (fast, capable) model from that same LLM provider for inline code completion. Let's create a little spec on how we would want to handle this before implementing it. 
3. Add support for Anthropic with the Anthropic SDK. This might be a bigger list if we have to abstract out all of the OpenAI types that we're currently relying on...

hegdes-usc commented on mito-ai/mito_ai/providers.py:
```diff
@@ -70,6 +70,7 @@ def __init__(self, **kwargs: Dict[str, Any]) -> None:
         super().__init__(log=get_logger(), **kwargs)
         self.last_error = None
         self._async_client: Optional[openai.AsyncOpenAI] = None
+        self._active_model = None
```
`constants.ACTIVE_MODEL` records the user's selection in the UI
`self._active_model` tracks what model the current client was built with

Removing `self._active_model` would eliminate the optimization that prevents rebuilding the client when unnecessary.

hegdes-usc commented on mito-ai/src/components/ModelSelector.tsx:
```diff
@@ -0,0 +1,91 @@
+    } else if (newModel.startsWith('gemini')) {
+      localStorage.setItem('GEMINI_MODEL', newModel);
+    } else if (newModel === 'ollama') {
+      localStorage.setItem('OLLAMA_MODEL', newModel);
```
Fixed in: [87c6f6f](https://github.com/mito-ds/mito/pull/1658/commits/87c6f6f45d72fc18d92db22cb949c8167ca4f906)

hegdes-usc commented on mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx:
```diff
@@ -140,6 +141,24 @@ const ChatTaskpane: React.FC<IChatTaskpaneProps> = ({
     const streamingContentRef = useRef<string>('');
     const streamHandlerRef = useRef<((sender: CompletionWebsocketClient, chunk: ICompletionStreamChunk) => void) | null>(null);
 
+    const updateModelOnBackend = async (model: string): Promise<void> => {
```
Fixed it via a change in the [mito-ai/src/components/ModelSelector.tsx](https://github.com/mito-ds/mito/pull/1658/files/a803de5c906607df4a5a5d2ab24ea9100f2ee90d#diff-f6fec345d523f09a1955ac6b5d58f9dd4b6e9d34b937e37b06c14b56a41d5e9d)

Commit id: [87c6f6f](https://github.com/mito-ds/mito/pull/1658/commits/87c6f6f45d72fc18d92db22cb949c8167ca4f906)

aarondr77 reviewed (commented):
Okay, very nice. We're so close to finishing this. I love the transition polishes you added when the dropdown resizes. Looks great! 

I found a few bugs that we need to address before we can merge it into dev. Let me know if you want to chat through any of them.

aarondr77 commented:
Oh let's also make sure that the `mito-ai src lint`, `mito-ai mypy`, and `mito ai backend` tests are passing.
